### PR TITLE
[ESI] Baseline XRT support

### PR DESCRIPTION
**Pull request by teqdruid**
Adds a "BSP" for XRT which just sets up the infrastructure. Outputs the "build package" complete with everything necessary to do an XRT image build and connect to the resulting accelerator. Does not have any communication with the accelerator. That's the next PR.

### CODE REVIEW

**Comment by teqdruid** [comment]
@jptech / @karakagi would you mind reviewing this?

File: frontends/PyCDE/integration_test/bsp/test_xrt.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+@module
+class Top:
+  clk = Clock(types.i1)
+  rst = Input(types.i1)
+
+  @generator
+  def construct(ports):
+    pass
+
+
```
Comment (prathik-sannecy-microsoft) [inline frontends/PyCDE/integration_test/bsp/test_xrt.py:22]: No input sanitization?

File: frontends/PyCDE/src/pycde/bsp/Makefile.xrt.j2 [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+	@echo "*************************"
+
+	az storage blob upload \
+		--subscription $(AZ_FPGA_SUB) \
+		--account-name $(AZ_FPGA_STORAGE_ACCOUNT) \
+		--container-name $(AZ_FPGA_STORAGE_CONTAINER) \
+	  --sas-token "$(SAS)" --overwrite \
+		--name $(IMAGE_AZ_NAME).xclbin --file $(NAME).hw.xclbin
+
+	bash validate-fpgaimage.sh --storage-account $(AZ_FPGA_STORAGE_ACCOUNT) \
```
Comment (prathik-sannecy-microsoft) [inline frontends/PyCDE/src/pycde/bsp/Makefile.xrt.j2:136]: Nit : formatting

File: frontends/PyCDE/integration_test/bsp/test_xrt.py (hunk 1 above)
Comment (teqdruid) [reply frontends/PyCDE/integration_test/bsp/test_xrt.py:22]: Nah. It's just a test.

Comment (toddmass) [reply frontends/PyCDE/integration_test/bsp/test_xrt.py:22]: How is this test run? Not pytest? Not part of a pipeline?
# RUN: %PYTHON% %s %t genhw 2>&1

This doesn't tell users what the arguments are and it's not hard to provide the help and checks on input arguments.

Comment (joerawson) [reply frontends/PyCDE/integration_test/bsp/test_xrt.py:22]: Speaking as a former QA engineer, writing undocumented crappy tests is a good way to make sure those tests don't get ran over time and get deleted.

File: frontends/PyCDE/src/pycde/bsp/Makefile.xrt.j2 [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+endif
+
+VPPFLAGS += -t $(TARGET) --platform $(PLATFORM)
+PACKAGE := $(BUILD)/package
+
+ifneq ($(TARGET), hw)
+VPPFLAGS += -g
+endif
+
+device2xsa = $(strip $(patsubst %.xpfm, % , $(shell basename $(PLATFORM))))
```
[the remark this answers is not in the document]
Comment (teqdruid) [reply frontends/PyCDE/src/pycde/bsp/Makefile.xrt.j2:52]: PLATFORM can be set when you run the Makefile. It's not resolvable at template generation time.

File: frontends/PyCDE/integration_test/bsp/test_xrt.py (hunk 1 above)
Comment (teqdruid) [reply frontends/PyCDE/integration_test/bsp/test_xrt.py:22]: It's a [`lit`](https://www.llvm.org/docs/CommandGuide/lit.html) test. Standard throughout LLVM/MLIR/CIRCT.

File: frontends/PyCDE/src/pycde/bsp/Makefile.xrt.j2 (hunk 3 above)
[the remark this answers is not in the document]
Comment (joerawson) [reply frontends/PyCDE/src/pycde/bsp/Makefile.xrt.j2:52]: So execution of the rendered makefile isn't tightly coupled with a synthesis, ok.

[the remark this answers is not in the document]
Comment (teqdruid) [reply frontends/PyCDE/src/pycde/bsp/Makefile.xrt.j2:52]: Yeah, this gets rendered into a 'build package' along with all the other build collateral. This enables users to run the heavy compiles anywhere just by copying the package dir. (As per https://github.com/llvm/circt/pull/4537/files#diff-4d9c5ea53a0073ec0a892d1493c09417cc106117ea67ef21c88f8e2787f37ad8R92).

File: frontends/PyCDE/src/pycde/system.py [hunk 4]
```diff
@@ -101,11 +101,24 @@ def add_packaging_step(self, func: Callable):
 
   @property
   def hw_output_dir(self):
-    return self.output_directory / "hw"
+    ret = self.output_directory / "hw"
+    if not ret.exists():
+      ret.mkdir()
```
Comment (toddmass) [inline frontends/PyCDE/src/pycde/system.py:106]: So if the hw directory has existing file in it? what about cleaning it?
This function name doesn't make it clear that any action is expected from calling this function to get the path to the hw directory.

File: frontends/PyCDE/src/pycde/system.py [hunk 5]
```diff
@@ -101,11 +101,24 @@ def add_packaging_step(self, func: Callable):
 
   @property
   def hw_output_dir(self):
-    return self.output_directory / "hw"
+    ret = self.output_directory / "hw"
```
Comment (toddmass) [inline frontends/PyCDE/src/pycde/system.py:104]: nit: I am used to a style that expects function calls to be self.output_directory() to distinguish from class properties/variable names.

File: frontends/PyCDE/src/pycde/system.py [hunk 6]
```diff
@@ -101,11 +101,24 @@ def add_packaging_step(self, func: Callable):
...
   def runtime_output_dir(self):
-    return self.output_directory / "runtime"
+    ret = self.output_directory / "runtime"
+    if not ret.exists():
+      ret.mkdir()
+    return ret
+
+  @property
+  def sys_runtime_output_dir(self):
+    ret = self.runtime_output_dir / self.name
```
Comment (toddmass) [inline frontends/PyCDE/src/pycde/system.py:118]: nit: this pattern is repeated 3 times for 3 different folder names. private internal function with string argument would standardize.

File: frontends/PyCDE/src/pycde/bsp/Makefile.xrt.j2 (hunk 2 above)
Comment (teqdruid) [reply frontends/PyCDE/src/pycde/bsp/Makefile.xrt.j2:136]: This is a result of tabs vs. spaces. I use two spaces for tabs whereas GH displays 4. Tabs... *sigh*.

File: frontends/PyCDE/src/pycde/system.py (hunk 4 above)
Comment (teqdruid) [reply frontends/PyCDE/src/pycde/system.py:106]: I'm not big on deleting files, especially considering that there could be a compiled image in the build package.

**Pull request merged**